meta-aggregator-contract/contracts/MetaAggregatorSwapContract.sol

```diff
@@ -0,0 +1,293 @@
+     * @param minAmountOut The minimum amount of tokenOut expected.
+     * @param receiver The address to receive the output tokens.
+     * @param isDelegate Whether to use delegatecall for the swap.
+     * @return amountOut The amount of tokenOut received.
+     */
+    function _swapETH(
+        IERC20 tokenIn,
+        IERC20 tokenOut,
+        address aggregator,
+        bytes calldata swapData,
+        uint256 amountIn,
+        uint256 minAmountOut,
+        address receiver,
+        bool isDelegate
+    ) internal returns (uint256 amountOut) {
+        if (receiver == address(0)) {
+            revert InvalidReceiver();
+        }
+        if (amountIn == 0) {
+            revert AmountInMustBeGreaterThanZero();
+        }
+        if (minAmountOut == 0) {
+            revert MinAmountOutMustBeGreaterThanZero();
+        }
+        if (address(tokenIn) == address(tokenOut)) {
+            revert TokenInAndTokenOutCannotBeSame();
+        }
+
+        if (msg.value != amountIn) {
+            revert IncorrectEtherAmountSent();
+        }
+
+        uint256 balanceBeforeSwap = tokenOut.balanceOf(address(this)); // Get the balance after the swap
+
+        _callAggregator(swapData, isDelegate, aggregator);
+
+        uint256 balanceAfterSwap = tokenOut.balanceOf(address(this));
+
+        amountOut = balanceAfterSwap - balanceBeforeSwap;
+        if (amountOut < minAmountOut) {
+            revert InsufficientOutputBalance();
+        }
+
+        TransferHelper.safeTransfer(address(tokenOut), receiver, amountOut);
+    }
+
+    /**
+     * @dev Internal function to perform the swap logic.
+     * @param tokenIn The input token (ERC20).
+     * @param tokenOut The output token (ERC20 or ETH).
+     * @param aggregator The address of the aggregator to perform the swap.
+     * @param swapData The data required for the swap.
+     * @param amountIn The amount of tokenIn to swap.
+     * @param minAmountOut The minimum amount of tokenOut expected.
+     * @param receiver The address to receive the output tokens.
+     * @param isDelegate Whether to use delegatecall for the swap.
+     * @return amountOut The amount of tokenOut received.
+     */
+    function _swapERC20(
+        IERC20 tokenIn,
+        IERC20 tokenOut,
+        address aggregator,
+        bytes calldata swapData,
+        uint256 amountIn,
+        uint256 minAmountOut,
+        address receiver,
+        bool isDelegate
+    ) internal returns (uint256 amountOut) {
+        if (receiver == address(0)) {
+            revert InvalidReceiver();
+        }
+        if (amountIn == 0) {
+            revert AmountInMustBeGreaterThanZero();
+        }
+        if (minAmountOut == 0) {
+            revert MinAmountOutMustBeGreaterThanZero();
+        }
+        if (address(tokenIn) == address(tokenOut)) {
+            revert TokenInAndTokenOutCannotBeSame();
+        }
+
+        TransferHelper.safeApprove(address(tokenIn), aggregator, 0);
```

Comment:
If we deploy to ETH we should add an if check here and only reset the approval for USDT

Reply:
@aj07 @Havoc19 what do you think? do we need to add the check here?